repondeur/zam_repondeur/fetch/an/dossiers/dossiers_legislatifs.py

```diff
@@ -212,45 +214,87 @@ def gen_lectures(
 
     def _walk_actes(acte: dict) -> Iterator[WalkResult]:
-        nonlocal current_texte
-
-        code = acte["codeActe"]
-        premiere_lecture = code.startswith("AN1") or code.startswith("SN1")
-        phase = code.split("-", 1)[1] if "-" in code else ""
-
-        if phase in {"COM-FOND", "COM-AVIS", "DEBATS"}:
-            if current_texte is not None:
-                yield WalkResult(
-                    phase=phase,
-                    organe=acte["organeRef"],
-                    texte=current_texte,
-                    premiere_lecture=premiere_lecture,
+        nonlocal texte_depose, texte_commission
+
+        chambre, phase, etape = parse_code_acte(acte["codeActe"])
+
+        if etape in {"COM-FOND", "COM-AVIS"}:
+            texte_examine = texte_depose
+        elif etape == "DEBATS":
+            texte_examine = texte_commission
+            if texte_commission is None:
+                logger.warning(
+                    f"{dossier_uid}->{acte['uid']}: pas de rapport de la commission"
+                    " saisie au fond, examen du texte déposé"
                 )
-            else:
-                logger.warning(f"Could not match a text for {acte['uid']}")
+                texte_examine = texte_depose
+        else:
+            texte_examine = None
 
-        for key in ["texteAssocie", "texteAdopte"]:
-            if key in acte and acte[key] is not None:
-                uid = acte[key]
-                if uid[:4] in {"PRJL", "PION"}:
-                    current_texte = uid
+        if texte_examine is not None:
+            yield WalkResult(
+                chambre=chambre,
+                phase=phase,
+                etape=etape,
+                organe=acte["organeRef"],
+                texte_examine=texte_examine,
+            )
+
+        # Texte déposé
+        if etape == "DEPOT":
+            texte_depose = acte["texteAssocie"]
+            texte_commission = None
+
+        # Texte adopté en commission (ou "null" si aucun amendement n'est adopté)
+        if etape == "COM-FOND-RAPPORT":
+            if acte["texteAdopte"] is not None:
+                texte_commission = acte["texteAdopte"]
+            else:
+                texte_commission = texte_depose
 
         for sous_acte in extract_actes(acte):
             yield from _walk_actes(sous_acte)
 
     yield from _walk_actes(acte)
 
 
+_CHAMBRES = {"AN": Chambre.AN, "SN": Chambre.SENAT}
+
+
+_PHASES = {
+    "1": Phase.PREMIERE_LECTURE,
+    "2": Phase.DEUXIEME_LECTURE,
+    "NLEC": Phase.NOUVELLE_LECTURE,
+    "LDEF": Phase.LECTURE_DEFINITIVE,
+}
+
+
+def parse_code_acte(code_acte: str) -> Tuple[Chambre, Phase, str]:
+    """
+    Phase: 1re lecture, 2e lecture, nouvelle lecture, lecture définitive
+    Étape: commission, séance, publique...
+    """
+    mo = re.match("(?P<chambre>AN|SN)(?P<phase>[A-Z0-9]+)(-(?P<etape>.*))?", code_acte)
+    if mo is None:
+        raise ValueError(f"Could not parse codeActe: {code_acte!r}")
```

Comment:
It might be useful to launch the `zam_fetch_amendements` command to check that we don't raise any error for existing lecture? 🤔

Reply:
The script looks kind of broken on master already, will fix in another PR.